compiler/codegen/OMRCodeGenerator.hpp

```diff
@@ -760,6 +759,28 @@ class OMR_EXTENSIBLE CodeGenerator
 
    uint32_t getPreJitMethodEntrySize() {return _preJitMethodEntrySize;}
    uint32_t setPreJitMethodEntrySize(uint32_t s) {return (_preJitMethodEntrySize = s);}
+   
+   /** \brief
+    *     Determines whether the code generator supports or allows JIT-to-JIT method entry point alignment.
+    */
+   bool supportsJitMethodEntryAlignment();
+
+   /** \brief
+    *     Determines the byte boundary at which to align the JIT-to-JIT method entry point. If the boundary is
+    *     specified to be \c x and the JIT-to-JIT method entry point to be \c y then <c>y & (x - 1) == 0</c>.
+    */
+   uint32_t getJitMethodEntryAlignmentBoundary();
+   
+   /** \brief
+    *     Determines the byte threshold at which the JIT-to-JIT method entry point boundary alignment will not be
+    *     performed. If the JIT-to-JIT method entry point is already close to the boundary then it may not make sense
+    *     to perform the boundary alignment as much code cache can be wasted. This threshold can be used to avoid such
+    *     situations.
+    *
+    *  \note
+    *     This value must be less than or equal to the boundary returned via \see getJitMethodEntryAlignmentBoundary.
+    */
+   uint32_t getJitMethodEntryAlignmentThreshold();
```

Comment:
While I understand the spirit of this option in reducing space, I'm not sure I understand the practical use of it.

Ultimately, aligning the entry point of a method on a particular boundary is a performance optimization.  However, the use of a threshold introduces non-determinancy (and arguably randomness run-to-run) into any performance benefit.  Whether or not a particular method is aligned depends on the methods that have been compiled before it.  It also doesn't take into account the "importance" of any particular method.  For instance, a method invoked millions of times is better suited to be aligned than one only invoked a few times, but it could be rejected for alignment because it exceeds the threshold.  Unless you align every method boundary I don't know how you would achieve consistent performance.

Perhaps the use of the threshold should be tied to an "optimize-for-space" option if one valued code cache size more than potential runtime performance benefits?

Reply:
I would tend to agree. This PR preserves functionality, and since this threshold existed on the Z codegen prior to this change we don't change that fact here. I too would like to remove the threshold as I'm never a fan of non-determinism and, to me personally I find it hard to imagine this threshold does anything to performance.

If you agree we can propose to remove the threshold altogether here after I carry out a performance experiment on Z to validate that not having the threshold will not impact performance. Please let me know whether I should proceed with this.

Reply:
I'm fine with discussing it here in this PR while you're actively modifying this code, or in a separate issue.  A separate issue might make it easier to find the discussion at a later date and explicitly focuses on deprecating the functionality.

Reply:
I've opened up #4373 to track this work. We can leave the API as is here and carry out the performance measurements in #4373 and deprecate in another PR.